task-maker-async/src/store.rs

```diff
@@ -0,0 +1,115 @@
+#![allow(dead_code)]
+
+use serde::{Deserialize, Serialize};
+
+type HashData = Vec<u8>;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct InputFileHash(HashData);
+
+/// Two-level hash; the outer hash has information about all properties of a computation that are
+/// expected to change the result (such as outer hashes of inputs, or the command line). The inner
+/// hash takes care of properties of the computation that should not change the outputs, such as
+/// time and memory limits; it also includes the first hash.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ComputationHash(HashData, HashData);
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct InputFileWriteHandle(usize);
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ComputationWriteHandle(usize);
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ReadHandle(usize);
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum ComputationDataFile {
+    Outcome,
+    OutputFile(String),
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum FileReadingOutcome {
+    /// The file has been deleted, for example because the worker responsible for the execution has
+    /// disappeared.
+    Dropped,
+    /// The file has been fully read.
+    EndOfFile,
+    /// A new chunk of data is available.
+    Data(Vec<u8>),
+}
+
+#[tarpc::service]
+pub trait Store {
+    /// Used by the client to upload inputs to the DAG. Returns None if the file is already
+    /// present. Either way, this leases the corresponding hash for LEASE_LENGTH_SEC.
+    async fn create_input_file(hash: InputFileHash) -> Result<InputFileWriteHandle, ()>;
+
+    /// Appends a chunk to the input file; the InputFileWriteHandle must have been obtained through
+    /// create_input_file. Renews the lease on the corresponding hash.
+    async fn append_input_chunk(file_handle: InputFileWriteHandle, data: Vec<u8>)
+        -> Result<(), ()>;
+
+    /// Finalizes the input file; no more data can be appended, and the resulting hash must match
+    /// the claimed hash when create_input_file was called.
+    async fn finalize_input_write_file(file_handle: InputFileWriteHandle) -> Result<(), ()>;
```

Comment:
What prevents the file from disappearing immediately after it's written (after all, no one's holding a read lease)?

Similarly for computation results.

Reply:
Also a fair point. One option could be that finalizing an input file actually converts your write lease/handle into a read lease/handle.

The same wouldn't work for computations, but there the problem doesn't quite exist as the server (and the client) will be waiting for the computation to finish and obtain a read lease to it as soon as it's finalized.

Reply:
Made this change for input files.